### PR TITLE
Fix encoding issues when printing/saving a form with non-ascii characters

### PR DESCRIPTION
**Pull request by calixteman**
This patch aims to fix issue #12280.
I played with this japanese form:
https://www.nta.go.jp/taxes/tetsuzuki/shinsei/annai/joyaku/annai/pdf2/250.pdf
I saved a modified version of this file and open it in Acrobat Reader (osx version) successfully.

@brendandahl, @timvandermeij, @Snuffleupagus I'm not completely sure that the fonction Font::encodeString is correct so don't hesitate to help me to fix it, your feedback is highly appreciated.

### CODE REVIEW

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
Really sorry, but I don't think that the changes you're making in `src/core/primitives.js` (and its usages) are at all correct or desirable here (it's simply *too* hacky in my opinion); this is really not how a `Dict` is supposed to work!

The correct solution might be to fix issue #12226 *first*, see the in-progress patch in PR #12207 (which I believe will require a fair number of changes before it's ready to land), since that being fixed *correctly* should mean that the `AcroForm` font-resources are available in a more standardized way.

(Also, merging of `Dict` data needs to be done *very* carefully if that's actually necessary.)

---

Other than that, I've not really looked at the rest of the patch since I believe that the above is enough of a blocker.

**Comment by calixteman** [comment]
Agreed, I forgot to move it in another patch.
Anyway the issue with this form is that the font we have in DA is /MSGothic and it isn't in DR but in Acroform::DR which is obviously an error.
I will fix this issue in another patch.

**Comment by calixteman** [comment]
I removed the DR issue fix so there is non-buggy textfield to play with: "Domicile or residence" one on the first page.

**Comment by calixteman** [comment]
I filed an issue for the DA/DR problem:
https://github.com/mozilla/pdf.js/issues/12294

**Comment by Snuffleupagus** [comment]
> I filed an issue for the DA/DR problem:
> #12294

But isn't that just another version of issue #12226, as mention above!?

**Comment by timvandermeij** [comment]
Overall this patch looks good to me, but I'm having trouble reviewing the `fonts.js` code since I'm not really familiar with it. Maybe @Snuffleupagus or @brendandahl can chime in for that?

File: src/core/annotation.js [hunk 1]
```diff
@@ -1286,37 +1297,42 @@ class WidgetAnnotation extends Annotation {
...
     const [font, fontName] = fontInfo;
-    const fontSize = this._computeFontSize(...fontInfo, totalHeight);
     this._fontName = fontName;
 
-    let descent = font.descent;
-    if (isNaN(descent)) {
-      descent = 0;
-    }
+    const lineHeight = totalHeight;
+    const fontSize = this._computeFontSize(...fontInfo, lineHeight);
```
Comment (Snuffleupagus) [inline src/core/annotation.js:1303]: The `lineHeight` doesn't seem to be used for anything else here; can you please revert these changes to reduce the overall size of the diff?

File: src/core/annotation.js [hunk 2]
```diff
@@ -1286,37 +1297,42 @@ class WidgetAnnotation extends Annotation {
...
+        totalHeight,
+        alignment,
         hPadding,
         vPadding
       );
     }
 
-    if (this.data.multiLine) {
-      return this._getMultilineAppearance(
+    // TODO: need to handle chars which are not in the font
```
Comment (Snuffleupagus) [inline src/core/annotation.js:1327]: Nit: Period at the end of sentences please!

File: src/core/annotation.js [hunk 3]
```diff
@@ -1568,49 +1601,59 @@ class TextWidgetAnnotation extends WidgetAnnotation {
   }
 
   _splitLine(line, font, fontSize, width) {
-    if (line.length <= 1) {
+    // TODO: need to handle chars which are not in the font
```
Comment (Snuffleupagus) [inline src/core/annotation.js:1604]: Nit: Period at the end of sentences please!

File: src/core/annotation.js [hunk 4]
```diff
@@ -1373,10 +1391,24 @@ class WidgetAnnotation extends Annotation {
...
-      // According to https://en.wikipedia.org/wiki/Em_(typography)
-      // an average cap height should be 70% of 1em
-      const capHeight = 0.7 * em;
+      let capHeight;
+      if (font.capHeight) {
+        capHeight = font.capHeight;
+      } else {
+        const glyphs = font.charsToGlyphs(
+          font.encodeString("M").join(""),
+          true
```
Comment (Snuffleupagus) [inline src/core/annotation.js:1400]: That method doesn't take more than one parameter, why are you passing `true` here!?

File: src/core/cmap.js [hunk 5]
```diff
@@ -338,6 +338,22 @@ class CMap {
...
+      const codespaceRange = codespaceRanges[n];
+      for (let k = 0, kk = codespaceRange.length; k < kk; ) {
+        const low = codespaceRange[k++];
+        const high = codespaceRange[k++];
+        if (charCode >= low && charCode <= high) {
+          return n + 1;
+        }
+      }
+    }
+    return 0;
```
Comment (Snuffleupagus) [inline src/core/cmap.js:354]: Is this default value correct, since the `readCharCode` method above returns `length = 1` in the fallback case?

File: src/core/fonts.js [hunk 6]
```diff
@@ -3351,9 +3352,119 @@ var Font = (function FontClosure() {
       return (charsCache[charsCacheKey] = glyphs);
     },
 
+    getGlyphsAndCharPosition: function Font_getGlyphsAndCharPosition(chars) {
```
Comment (Snuffleupagus) [inline src/core/fonts.js:3355]: Nit: Here and elsewhere below, when adding *new* code you can simply format methods as `getGlyphsAndCharPosition(chars) {` instead.

File: src/core/fonts.js [hunk 7]
```diff
@@ -3351,9 +3352,119 @@ var Font = (function FontClosure() {
...
+        }
+      }
+
+      return glyphs;
+    },
+
+    getCharPosition: function Font_getCharPosition(chars) {
+      // Chars can have different sizes
+      // so get the char positions in the string
+      // it's useful to split a string at correct char positions.
```
Comment (Snuffleupagus) [inline src/core/fonts.js:3391]: This isn't all that easy to ready, since it feels like you're using line-breaks instead of proper punctuation here.
Also, should this information perhaps be a JSDoc comment instead?

File: src/core/fonts.js [hunk 8]
```diff
@@ -3351,9 +3352,119 @@ var Font = (function FontClosure() {
...
+    },
+
     get glyphCacheValues() {
       return Object.values(this.glyphCache);
     },
+
+    encodeString: function Font_encodeString(str) {
+      // The function returns an array of strings
+      // at even positions there is a valid string (may be empty)
+      // and at odd positions string where chars are not in the font
```
Comment (Snuffleupagus) [inline src/core/fonts.js:3419]: Again this isn't all that easy to ready, since it feels like you're using line-breaks instead of proper Spunctuation here.
Also, should this information perhaps be a JSDoc comment instead?

File: src/core/fonts.js [hunk 9]
```diff
@@ -3351,9 +3352,119 @@ var Font = (function FontClosure() {
...
+      const buffers = [];
+      let buf = [];
+      let isError = false;
+
+      for (let i = 0, ii = str.length; i < ii; i++) {
+        const unicode = str.codePointAt(i);
+        let hasError = false;
+        if (unicode > 0xd7ff && (unicode < 0xe000 || unicode > 0xfffd)) {
+          // unicode is represented by two u16
```
Comment (Snuffleupagus) [inline src/core/fonts.js:3428]: What is `u16`? Please improve/expand this comment, since it's not all that clear currently!

File: src/core/fonts.js (hunk 6 above)
Comment (Snuffleupagus) [reply src/core/fonts.js:3355]: Also, this is an *almost* verbatim copy of the existing `charsToGlyphs` method, which feels quite unfortunate!
Furthermore, this ignores the existing caching that's used in that method, which can lead to quite a bit of unnecessary re-parsing here.

Given how this is used, with only one call-site as far as I can tell, could you not simply call both the *existing* `charsToGlyphs` and then the *new* `getCharPosition` methods in place of this "combine" one!?
(Considering the current lack of caching in the new methods, I really don't see that being much slower overall.)

File: src/core/fonts.js (hunk 7 above)
Comment (Snuffleupagus) [reply src/core/fonts.js:3391]: Also, you probably want to add a *lazily* initialized cache for the data (something similar to `charsToGlyphs` above), to avoid unnecessary re-parsing here.

File: src/core/fonts.js [hunk 10]
```diff
@@ -3351,9 +3352,119 @@ var Font = (function FontClosure() {
...
+        }
+        if (this.toUnicode) {
+          const char = String.fromCodePoint(unicode);
+          const charCode = this.toUnicode.charCodeOf(char);
+          if (charCode === -1) {
+            hasError = true;
+          } else {
+            if (isError) {
+              buffers.push(buf.join(""));
+              buf = [];
```
Comment (Snuffleupagus) [inline src/core/fonts.js:3439]: By doing `buf.length = 0`, you can re-use the *same* Array rather than creating a new one (which should be more efficient overall).

File: src/core/fonts.js [hunk 11]
```diff
@@ -3351,9 +3352,119 @@ var Font = (function FontClosure() {
...
       return Object.values(this.glyphCache);
     },
+
+    encodeString: function Font_encodeString(str) {
+      // The function returns an array of strings
+      // at even positions there is a valid string (may be empty)
+      // and at odd positions string where chars are not in the font
+      const buffers = [];
+      let buf = [];
+      let isError = false;
```
Comment (Snuffleupagus) [inline src/core/fonts.js:3422]: The very similarly named `isError`/`hasError` variables makes this code, at first sight, a bit difficult to follow efficiently.

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]
I've added a couple of comments, based on an very cursory look at the patch.

File: src/core/fonts.js (hunk 6 above)
Comment (calixteman) [reply src/core/fonts.js]: You're right: I removed that stuff.

File: src/core/fonts.js [hunk 12]
```diff
@@ -3351,9 +3352,88 @@ var Font = (function FontClosure() {
...
+     * Encode a js string using font encoding.
+     * The resulting array contains an encoded string at even positions
+     * (can be empty) and a non-encoded one at odd positions.
+     * @param {String} a js string.
+     * @returns {Array<String>} an array of encoded strings or non-encoded ones.
+     */
+    encodeString(str) {
+      const buffers = [];
+      const currentBuf = [];
+      const hasCurrentBufErrors = () => buffers.length % 2 === 1;
```
Comment (Snuffleupagus) [inline src/core/fonts.js:3397]: The name of this helper function suggests (to me) that the intention is to check the `currentBuf` array, but that's not actually happening here; if this is indeed intended behaviour a comment might not hurt here!

**Review by Snuffleupagus — COMMENTED** [review]
Also, looking at the previous test-results there's some *vertical* movement of text in existing `print`-tests; was this an intentional change here?

**Review by Snuffleupagus — APPROVED** [review]
I've not looked in detail at the unit-tests, but overall the `core` (and importantly `fonts.js`) parts seem reasonable now :-)

I don't know if either Brendan or Tim wants to give this a final once-over, before we land it?

**Review by timvandermeij — APPROVED** [review]

**Comment by timvandermeij** [comment]
Looks good to me too. Thanks!
